### PR TITLE
[4.0] - Postgresql save article

### PR DESCRIPTION
**Pull request by alikon**
Pull Request for Issue: save article fail on pdo postgresql

### Summary of Changes
used proper UPDATE with JOIN syntax 

### Testing Instructions
Create an Article

### Expected result
Article created 

### Actual result
on Postgresql you got an ERROR
Screenshot from 2019-03-10 10-46-31

### CODE REVIEW

File: libraries/src/Table/Table.php [hunk 1]
```diff
@@ -1588,7 +1588,19 @@ public function reorder($where = '')
...
-		$query->innerJoin('(' . (string) $subquery . ') AS sq ON ' . implode(' AND ', $innerOn));
+		// Mysql diverge from Postgresql when UPDATE with JOIN
+		if ($this->_db->name === 'mysql')
+		{
+			$query->innerJoin('(' . (string) $subquery . ') AS sq ON ' . implode(' AND ', $innerOn));
+		}
+		else
+		{
+			$query->innerJoin('(' . (string) $subquery . ') AS sq ' );
+			foreach ($innerOn as $key)
```
Comment (Quy) [inline libraries/src/Table/Table.php]:
```suggestion

			foreach ($innerOn as $key)
```

File: libraries/src/Table/Table.php [hunk 2]
```diff
@@ -1588,7 +1588,20 @@ public function reorder($where = '')
...
 
-		$query->innerJoin('(' . (string) $subquery . ') AS sq ON ' . implode(' AND ', $innerOn));
+		// Mysql diverge from Postgresql when UPDATE with JOIN
+		if ($this->_db->name === 'mysql')
+		{
+			$query->innerJoin('(' . (string) $subquery . ') AS sq ON ' . implode(' AND ', $innerOn));
+		}
+		else
+		{
+			$query->innerJoin('(' . (string) $subquery . ') AS sq ' );
```
Comment (SharkyKZ) [inline libraries/src/Table/Table.php]:
```suggestion
			$query->innerJoin('(' . (string) $subquery . ') AS sq ');
```

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 4c65881eea0a0d388f7aa0793c7c88221fdbf980<br /><br />Was not easy to test because
1. Had to apply PR #24747 to be able to have installation on Postgresql for the test.
2. Patchtester does not work with Postgresql
3. Had to apply changes of this PR manually because branch or this PR needs a rebase to current 4.0-dev.
But at the end test has shown that this PR here corrects error as described.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24148">issues.joomla.org/tracker/joomla-cms/24148</a>.</sub>

**Comment by twister65** [comment]
I have tested this item :white_check_mark: successfully on 4c65881eea0a0d388f7aa0793c7c88221fdbf980<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24148">issues.joomla.org/tracker/joomla-cms/24148</a>.</sub>

**Comment by twister65** [comment]
@richard67 , I have modified patchtester to work with pgsql in J4:
[com_patchtester.zip](https://github.com/joomla/joomla-cms/files/3124978/com_patchtester.zip)

I make a PR for the next release:
https://github.com/joomla-extensions/patchtester/pull/220

**Comment by ghost** [comment]
Status "Ready To Commit".

**Comment by wilsonge** [comment]
This is the exact same line as in staging. Why is this an issue in J4 and not in staging :/

**Comment by alikon** [comment]
don't forget we use PDO postgresql in 4

**Comment by wilsonge** [comment]
This https://stackoverflow.com/questions/31740105/select-with-inner-join-using-postgresql-and-pdo suggests that PDO supports `ON` as a keyword. And using the second param of the `join` function allows specifying an `ON` condition. So we need to dig a bit further into this and figure out exactly where we are failing here because it must be something more specific than just an `ON` statement https://github.com/joomla-framework/database/blob/2.0-dev/src/DatabaseQuery.php#L1168

**Comment by alikon** [comment]
so let me clarify a  bit:

the query in staging 3.x is quite different from what we have here in 4.x

this is 3.x

```sql
UPDATE tre_content
INNER JOIN (
     SELECT * FROM (
         SELECT (
             SELECT @rownum := @rownum + 1 FROM (SELECT @rownum := 0) AS r) AS new_ordering,`id` AS `pk__0`
FROM tre_content
WHERE catid = 2 AND state >= 0 AND `ordering` >= 0
ORDER BY `ordering`
) w) AS sq ON `id` = sq.`pk__0`
SET `ordering` = sq.new_ordering
WHERE catid = 2 AND state >= 0 AND `ordering` >= 0
```

this is 4.x before patch

```sql
UPDATE quattro_content
SET "ordering" = sq.new_ordering
FROM ( SELECT ROW_NUMBER() OVER (ORDER BY "ordering") AS new_ordering,"id" AS "pk__0"
	FROM quattro_content
	WHERE catid = 2 AND state >= 0 AND "ordering" >= 0) AS sq ON "id" = sq."pk__0"
WHERE catid = 2 AND state >= 0 AND "ordering" >= 0"
```

this is 4.x with patch

```sql
UPDATE quattro_content
SET "ordering" = sq.new_ordering
FROM (
SELECT ROW_NUMBER() OVER (ORDER BY "ordering") AS new_ordering,"id" AS "pk__0"
FROM quattro_content
WHERE catid = 2 AND state >= 0 AND "ordering" >= 0) AS sq 
WHERE catid = 2 AND state >= 0 AND "ordering" >= 0 AND "id" = sq."pk__0"
```

so the query in 4.x is using and update with join on a temporay table (look at the `FROM` clause) and doesn't like the `ON` clause instead need the `AND` clause

**Comment by wilsonge** [comment]
That seems very weird. How come in 4.x we don't have an inner join in the query at all. Given that's the command we're modifying in the code. Are you sure that's the right query?

**Comment by alikon** [comment]
pretty sure i've made my homework, but another couple of eyes checking it's always better

**Comment by richard67** [comment]
@alikon Last bigger change on that query was PR #13505 . This is shown in the history and blame for that file on both the staging and the 4.0-dev branch. And when comparing the sql statement related code on function reorder between 4.0-dev and staging, I also see no difference, only differences to other stuff. When you checked 3.x, did you maybe check something older than 3.7?

**Comment by alikon** [comment]
from 3.x to 4.x something has changed :smiley: .... maybe the framework ... don't know,
btw, i've already seen some strange behavior with pdo pgsql

**Comment by richard67** [comment]
@alikon Hmm, I just see that with our PR applied, the variable `$subquery` will not be used anymore in that function. So either it has to be removed, or there is really something wrong.

**Comment by alikon** [comment]
it is used here https://github.com/joomla/joomla-cms/pull/24148/files#diff-f70cde2615a113ab8ff64af30ef43f5fR1603

**Comment by richard67** [comment]
@alikon Stupid me ;-)

**Comment by richard67** [comment]
Query in 4.0-dev **without** this PR, **MySQL**:

> UPDATE j4ux0_content
> INNER JOIN (
> SELECT * FROM (
> SELECT (SELECT @rownum := @rownum + 1 FROM (SELECT @rownum := 0) AS r) AS new_ordering,`id` AS `pk__0`
> FROM j4ux0_content
> WHERE catid = 2 AND state >= 0 AND `ordering` >= 0
> ORDER BY `ordering`
> ) w) AS sq ON `id` = sq.`pk__0`
> SET `ordering` = sq.new_ordering
> WHERE catid = 2 AND state >= 0 AND `ordering` >= 0

Query in 4.0-dev **with** this PR, **MySQL**:

> UPDATE j4ux0_content
> INNER JOIN (
> SELECT * FROM (
> SELECT (SELECT @rownum := @rownum + 1 FROM (SELECT @rownum := 0) AS r) AS new_ordering,`id` AS `pk__0`
> FROM j4ux0_content
> WHERE catid = 2 AND state >= 0 AND `ordering` >= 0
> ORDER BY `ordering`
> ) w) AS sq 
> SET `ordering` = sq.new_ordering
> WHERE catid = 2 AND state >= 0 AND `ordering` >= 0 AND `id` = sq.`pk__0`

I will later post the same for PDO **PgSQL**.

Output produced with `error_log($query->dump());` before the `$this->_db->setQuery($query);`.

**Comment by richard67** [comment]
Query in 4.0-dev **without** this PR, **PDO PgSQL**:

> UPDATE j4ux0_content
> SET "ordering" = sq.new_ordering
> FROM (
> SELECT ROW_NUMBER() OVER (ORDER BY "ordering") AS new_ordering,"id" AS "pk__0"
> FROM j4ux0_content
> WHERE catid = 2 AND state >= 0 AND "ordering" >= 0) AS sq ON "id" = sq."pk__0"
> WHERE catid = 2 AND state >= 0 AND "ordering" >= 0

=> It seems the join is changed to something else for PDO PgSQL, and without a join, any "ON" is a syntax error.

So the question is what removes that join. Is there something wrong with the db driver? This has to be fixed, and then this PR here might be obsolete.

**Comment by richard67** [comment]
Also the syntax `UPDATE table SET column=value FROM ( ... )  AS x ON y ... ` seems wrong to me. I've never heard about `UPDATE ... FROM ...`, I only know `INSERT INTO ... SELECT FROM ...`.

**Comment by richard67** [comment]
Maybe "PDO" just needs to be renamed to "PDU", "Please Don't Use". ;-)

**Comment by richard67** [comment]
I've just checked that the queries with MySQL (PDO) are the same as with MySQLi. So the problem with the join being changed to something else appears only with PostgreSQL (PDO).

**Comment by twister65** [comment]
https://www.postgresql.org/docs/current/sql-update.html
RTC.

**Comment by twister65** [comment]
You are right @richard67, there is not `from` with MySQL:
https://dev.mysql.com/doc/refman/8.0/en/update.html

**Comment by wilsonge** [comment]
Slightly reluctantly merging this. Still seems odd to me. But if testing proves it ok. Will do for now

**Pull request merged**